File: Kattis/baconeggsandspam.cpp
/*
Problem: Bacon, Eggs, and Spam
URL: https://open.kattis.com/problems/baconeggsandspam
Language: C++
Author: dvchinx
*/

#include <bits/stdc++.h>
using namespace std;

vector<string> split(const string& texto) {
    stringstream ss(texto);
    string palabra;
    vector<string> resultado;

    while (ss >> palabra) 
        resultado.push_back(palabra);
    return resultado;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    while (true) {
        int N; cin >> N;
        if (N == 0) break;

        // Para almacenar de forma alfabetica alimentos y nombres
        // Se puede optimizar usando 'unordered' y ordenando antes de imprimir
        map<string, set<string>> report;

        cin.ignore();
        while (N--) {
            string name; cin >> name;
            string order; getline(cin, order);
            vector<string> orders = split(order);

            for (string o : orders) {
                // Agregar las comidas al mapa
                report[o].insert(name);
            }
        }
        
        for (auto it : report) {
            cout << it.first << " ";
            for (string names : it.second) {
                cout << names << " ";
            }
            cout << "\n";
        }
        cout << "\n";
    }

    return 0;
}

File: Kattis/baconeggsandspam.py
"""
Problem: Bacon, Eggs, and Spam
URL: https://open.kattis.com/problems/baconeggsandspam
Language: Python
Author: dvchinx
"""

import sys

def main():
    lines = sys.stdin.read().split("\n")
    idx = 0
    out = []

    while True:
        n = int(lines[idx]); idx += 1
        if n == 0:
            break

        report = {}
        for _ in range(n):
            parts = lines[idx].split(); idx += 1
            name, foods = parts[0], parts[1:]
            for food in foods:
                report.setdefault(food, set()).add(name)

        for food in sorted(report):
            out.append(food + " " + " ".join(sorted(report[food])))
        out.append("")

    print("\n".join(out))


if __name__ == "__main__":
    main()
